### PR TITLE
Refine host name error checks

### PR DESCRIPTION
**Pull request by martinpitt**
Follow-ups from https://github.com/allisonkarlitskaya/ferny/pull/25#discussion_r1439738653

### CODE REVIEW

File: src/ferny/ssh_errors.py [hunk 1]
```diff
@@ -41,6 +41,10 @@ def __init__(self, match: Match, stderr: str) -> None:
         self.message = match.group(0)
 
 
+class SshInvalidHostnameError(SshError):
+    PATTERN = re.compile(r'^hostname contains invalid characters', re.I)
```
Comment (martinpitt) [inline src/ferny/ssh_errors.py:45]: I deliberately did *not* match a `$` here -- future versions may well extend the error message to show the invalid character or some additional help. This is the most robust compromise IMHO.

Comment (allisonkarlitskaya) [reply src/ferny/ssh_errors.py:45]: This needs a testcase added in the `STDERR_MESSAGES` section of `test_regexp.py`

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
Thanks for acting on these.  See below.

File: test/test_basic.py [hunk 2]
```diff
@@ -94,9 +94,14 @@ async def test_connection_refused(runtime_dir: pathlib.Path) -> None:
 @pytest.mark.asyncio
 async def test_dns_error(runtime_dir: pathlib.Path) -> None:
     session = ferny.Session()
-    with pytest.raises((socket.gaierror, ferny.ssh_errors.SshError)):
+    # OpenSSH 0.9.6+ statically checks the name, earlier versions throw anything at the DNS resolver
+    with pytest.raises((socket.gaierror, ferny.ssh_errors.SshInvalidHostnameError)):
         await session.connect('¡invalid hostname!')
 
+    # this is a validly formatted hostname
```
Comment (allisonkarlitskaya) [inline test/test_basic.py:101]: I think this is a separate test case.

File: src/ferny/ssh_errors.py (hunk 1 above)
Comment (martinpitt) [reply src/ferny/ssh_errors.py:45]: Done.

File: test/test_basic.py (hunk 2 above)
Comment (martinpitt) [reply test/test_basic.py:104]: ok -- done.

**Review by allisonkarlitskaya — APPROVED** [review]
Thanks

**Pull request merged**